lib/src/package.dart

```diff
@@ -278,27 +278,57 @@ See $workspacesDocUrl for more information.
       return p.join(root, path);
     }
 
-    return Ignore.listFiles(
+    /// Throws if [path] is a link that cannot resolve.
+    ///
+    /// Circular links will fail to resolve at some depth defined by the os.
+    void verifyLink(String path) {
+      final link = Link(path);
+      if (link.existsSync()) {
+        try {
+          link.resolveSymbolicLinksSync();
+        } on FileSystemException catch (e) {
+          throw DataException(
+            'Could not resolve symbolic link $path. $e',
+          );
+        }
+      }
+    }
+
+    /// We check each directory that it doesn't symlink-resolve to the
+    /// symlink-resolution of any parent directory of itself. This avoids
+    /// cycles.
+    ///
+    /// Cache the symlink resolutions here.
+    final symlinkResolvedDirs = <String, String>{};
```

Comment:
```suggestion
    final symlinkResolvedDirs = <String, String>{};
    String resolveSymlink(String path) =>
      symlinkResolvedDirs[path] ??= Directory(path).resolveSymbolicLinksSync()
```

That might make the code below a bit more readable.